Added process metrics and refactored code of previous ones

stefanodallapalma opened the pull request:
Added metrics to:

- count the number of developers who modified the file during the prior release,
- count the number of new developers who modified the file during the prior release,
- count the normalized number (by the total number of added and deleted lines) of added and deleted lines.

stefanodallapalma commented:
Closed. Need to update the implementation of metrics to be in the scope of a single release.

ishepard commented:
Hi Stefano! Thank you for the PR.

One small suggestions: while for the first 2 process metrics the usage was clear, I don't see the purpose of these new process metrics. They seem ad-hoc features that no-one will ever use. 

For example, the difference between `developers_dev_count` and and "developers who modified the file during the prior release" is just that it's between 2 releases, instead of between 2 commits. I'd avoid creating this function, and let the user pass the 2 release commits to `developers_dev_count. Same thing for "new". 

Furthermore, I don't see these metrics in the paper.

stefanodallapalma commented:
> Hi Stefano! Thank you for the PR.
> 
> One small suggestions: while for the first 2 process metrics the usage was clear, I don't see the purpose of these new process metrics. They seem ad-hoc features that no-one will ever use.
> 
> For example, the difference between `developers_dev_count` and and "developers who modified the file during the prior release" is just that it's between 2 releases, instead of between 2 commits. I'd avoid creating this function, and let the user pass the 2 release commits to `developers_dev_count. Same thing for "new".
> 
> Furthermore, I don't see these metrics in the paper.

Hi Davide,
the paper is https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=4273264. It was cited in the previous one.

However, I am implementing new metrics, and also refactoring the methods into proper classes for each metric. Threat them as single methods in the class ProcessMetrics is a bit messy, as it is becoming a blob class.

Create a class for each metric would also make the code more comprehensible and easy to update.

Stay tuned :)

stefanodallapalma commented:
The PR contains the following metrics:

* Commits count: to count for the number of commits that touched a given file in the scope of a given release (the one containing the commit to analyse)

* Active and Cumulative Distinct Devs count: to count for the number of active and distinct developers who modified the file. Active developers modified the file in commits within the same release the file's commit belongs to; the cumulative number of developers is the total number of developers that modified the file up to the given file's commit.

* Normalized Lines Added/Deleted: to count for number of normalized (by the total number of added and deleted lines) and deleted lines in the file.

## Refactoring
### Defined class for each metric
Each metric is mapped to a class. This is to make the code more comprehensible and easy to update. Each class is a child of the main class for process metrics ```ProcessMetric```, which define only one method ```count()``` implemented by the actual classes defining the metrics.

### Added interface to call the metrics
A module ```pydriller/metrics/process/metrics.py``` has been created as a main interface to call the metrics.

### Test cases using unittest and parameterized
Test cases have been created using the framework ```unittest``` and the module ```parameterized``` to avoid duplicated code and make code more readable.
The module parameterized allow to create a single class to test a functionality. Test cases are expressed as a dictionary. Therefore, we will have a dictionary with "n" lines (each containing test data and expected value) instead of having "n" methods testing the same functionality with different data but with duplicated code.

ishepard commented:
Hi @stefanodallapalma! 
I will have a look at it soon. In these days I am busy for a submission. I just skimmed the PR and it looks good to me. 

Small thing: can you use pytest instead of unittest? For consistency reasons. There is a way to have parameters there as well, you can have a look at my other tests where I do it. 

I checked the error and it's the same error we were facing some time ago..I really don't know how to solve it. I will have to get a windows machine and spend some time with it.

ishepard commented:
@stefanodallapalma there seems to be another error on Windows now:
```
File "D:\a\pydriller\pydriller\tests\metrics\process\test_commits_count.py", line 22

       assert count == expected, f'Test failed because expected {str(expected)} and got {str(count)}!'

                                                                                                     ^
SyntaxError: invalid syntax
```

To be honest, since pytest prints expected and actual, you can remove the string! Unless you want to put a specific message.

I will try to have a look at the Windows error soon

stefanodallapalma commented:
> @stefanodallapalma there seems to be another error on Windows now:
> 
> ```
> File "D:\a\pydriller\pydriller\tests\metrics\process\test_commits_count.py", line 22
> 
>        assert count == expected, f'Test failed because expected {str(expected)} and got {str(count)}!'
> 
>                                                                                                      ^
> SyntaxError: invalid syntax
> ```
> 
> To be honest, since pytest prints expected and actual, you can remove the string! Unless you want to put a specific message.
> 

I will remove the messages in the next commit.

> I will try to have a look at the Windows error soon
Cool, let me know.

ishepard commented:
@stefanodallapalma I just pushed a change in master, I put appveyor back to test on windows. Could you please merge it on your branch? I would like to see if we have the same Windows problem there.

ishepard commented:
Ok so it fails even on Appveyor. It's not a problem of GitHub workflow.

ishepard commented:
It's really hard to understand what the problem is if I can't reproduce it 😠 
I just run your PR on my Windows machine and it works. 

Like we (SWEs) like to say: it works on my machine!!

ishepard commented:
Oh yyeesss, I discovered the bug. Just as a reference for the future:

As explained in the issue gitpython-developers/GitPython#546, in Windows there seems to be a problem when creating a repository as a context manager. If this happens, then the repo can not be deleted.

My guess is that we obtained the error **OSError: [WinError 6] The handle is invalid** because we were creating too many repository, keeping the reference, and at one point we could not open a `subprocess.Popen` anymore.

We fixed the problem by calling `clear_cache()` before moving to the next repository.

stefanodallapalma commented:
> Oh yyeesss, I discovered the bug. Just as a reference for the future:
> 
> As explained in the issue [gitpython-developers/GitPython#546](https://github.com/gitpython-developers/GitPython/issues/546), in Windows there seems to be a problem when creating a repository as a context manager. If this happens, then the repo can not be deleted.
> 
> My guess is that we obtained the error **OSError: [WinError 6] The handle is invalid** because we were creating too many repository, keeping the reference, and at one point we could not open a `subprocess.Popen` anymore.
> 
> We fixed the problem by calling `clear_cache()` before moving to the next repository.

17 tests failed because of those issues on my Windows machine. Glad you've already found and solved the problem!

ishepard commented on pydriller/metrics/process/new_devs_count.py:
```diff
@@ -0,0 +1,87 @@
+from pydriller.repository_mining import RepositoryMining
+from pydriller.metrics.process.process_metric import ProcessMetric
+
+class NewDevsCount(ProcessMetric):
```
Why is this not together with the other DevsCount?

stefanodallapalma replied:
It was the last metric I have implemented, and done in a different file for simplicity and to test. I will merge them.

ishepard replied:
perfect 👍

ishepard commented:
Couple of points:

1. it's quite hard for me to understand all these metrics. I never heard most of them, so I will trust you on maintaining them 😄 
2. there are some bugs here and there, some `break` that shouldn't be there, or wrong parameters. Check those and be careful
3. Overall, this PR is huge. 24 files changed is more than I can review in a single session. For the next time, please keep the PR small (<5 files). For example, by splitting it in multiple PRs (you could do a PR for each metric). In this way it's easier to review.
4. Overall, I think the structure looks ok! The only problem I have with this PR is the "calculate at release". It makes the code more complicated for us to maintain, and it doesn't provide any advantage to the user. If you need to count at release level, it's better to do:

```
releases_commits = get_releases()
for x in range(releases_commits, len(releases_commits) - 1):
    countsomething(from_commit= releases_commits[x],to_commit= releases_commits[x+1])
    # do something with your count, like saving it 
```
I think this would be better for the user, as well for us. So I think you should remove the `self.releases` filter.

stefanodallapalma commented on pydriller/metrics/process/hunks_count.py:
```diff
@@ -0,0 +1,45 @@
+        filepath = self.filepath
+
+        for commit in RepositoryMining(path_to_repo=self.path_to_repo,
+                                       single=self.to_commit).traverse_commits():
```
The number of hunks is measured for a single commit, there is no need to get the list of commits up to ```to_commit```

ishepard replied:
mmm, so hunk is calculated just in a single commit. then ignore my suggestion

stefanodallapalma commented on pydriller/metrics/process/commit_count.py:
```diff
@@ -0,0 +1,40 @@
+
+                    break
+
+            if self.release_scope and commit.hash in self.releases:
```
I meant the metrics to be release-duration, as described in the paper.
```self.releases``` in the ```ProcessMetrics``` class contains all the tagged commits for a given repo.  
It is used to find the release a given commit (to_commit) belongs to and to stop the count of the metric.

Imagine we want to calculate the metric for commit c6 below (where R1 and R2 are the tagged commits c1 and c4, respectively):

|---c2---c3---|---c5---c6---
R1-----------R2

The only commits to take into account are c4, c5 and c6.
Thus, the function ```traverse_commits()``` will yield c6, c5, c4, c3, c2, c1, but the metric count will stop at c4.

If you prefer to stop at ```from_commit``` and let decide the user the temporal range for the analysis, I will change it.

Honestly, I am now realizing that lot of projects do not tag commits, and this would limit the usefulness of the tool.

ishepard replied:
|---c2---c3---|---c5---c6
R1-----------R2

Ok. To me, these are the possible scenarios:
1. the user wants to run the analysis **between R1 and R2**: in this case, the user would have to know what commits are R1 and R2; hence, our `self.releases` is useless because the user already knows the commits
2. the user wants to run the analysis **from c6 to R2**: the user doesn't know the commit of R2, hence he relies on our `self.releases`. However, complicating the code for just this case is not worth it. The user can easily calculate the `to_commit` by himself 

hence, I suggest to just delete the release. I can do it by myself later when I merge the PR

stefanodallapalma replied:
ok, let's go for scenario 1. I will update the code. no worries :)

stefanodallapalma commented on pydriller/metrics/process/devs_count.py:
```diff
@@ -0,0 +1,58 @@
+                    if modified_file.change_type == ModificationType.RENAME:
+                        filepath = str(Path(modified_file.old_path))
+
+                    break
```
The metric is relative to a single file. Once the file is found among the modified files, then the code iterates to the next commit.
Otherwise, we can follow Marco's approach of using and returning a dictionary ```{filepath: count}```

ishepard replied:
I see, you want to stop when you meet the file. That's ok.
As I side note, consider that when I run `git diff` I get the diff of all the files contemporary. So breaking when you get the file won't change the runtime complexity.

stefanodallapalma commented on pydriller/metrics/process/lines_count.py:
```diff
@@ -0,0 +1,73 @@
+                    if modified_file.change_type == ModificationType.RENAME:
+                        filepath = str(Path(modified_file.old_path))
+
+                    break
```
See previous comment on break

stefanodallapalma commented:
> Couple of points:
> 
> 1. it's quite hard for me to understand all these metrics. I never heard most of them, so I will trust you on maintaining them smile

For me is the same for some of them. However, I just rely on the paper that I've shown you and its cited papers.
If you have some idea about other metrics that you know, please let me know.

> 2. there are some bugs here and there, some `break` that shouldn't be there, or wrong parameters. Check those and be careful

Yep

> 3. Overall, this PR is huge. 24 files changed is more than I can review in a single session. For the next time, please keep the PR small (<5 files). For example, by splitting it in multiple PRs (you could do a PR for each metric). In this way it's easier to review.

Right. After the PR I've started coding quite fast and commit often, which have been automatically added to the PR.

> 4. Overall, I think the structure looks ok! The only problem I have with this PR is the "calculate at release". It makes the code more complicated for us to maintain, and it doesn't provide any advantage to the user. If you need to count at release level, it's better to do:
> 
> ```
> releases_commits = get_releases()
> for x in range(releases_commits, len(releases_commits) - 1):
>     countsomething(from_commit= releases_commits[x],to_commit= releases_commits[x+1])
>     # do something with your count, like saving it 
> ```
> 
> I think this would be better for the user, as well for us. So I think you should remove the `self.releases` filter.

Got it. I will make the code simpler.

Thank you!

stefanodallapalma commented:
Ciao @ishepard, I have changed the code as follows:

- each metric requires an (optional) ```from_commit``` instead of working with ```self.releases```. If the user needs to compute a metric within the scope of a release, then he has to provide a ```from_commit``` and a ```to_commit``` that identify the release period.

- each metric is computed for every modified file (as previously suggested by Marco) in the repository within the considered evolution period ```[from_commit, to_commit]```. Therefore the results of each metric is a dictionary ```{filepath_i: metric_value_i, filepath_j: metric_value_j ... }```.

stefanodallapalma commented on pydriller/metrics/process/contributors_experience.py:
```diff
@@ -0,0 +1,53 @@
+            else:
+                files[path] = round(100*max(contributions.values()) / total, 2)
+
+        return files
```
This metric shares almost all of its body with the file ```contributors_count.py```, but the semantic is different and more related to the number of modified lines (as in ```lines_count.py```). That's why I have created a different class.

ishepard commented on tests/metrics/process/test_hunks_count.py:
```diff
@@ -0,0 +1,18 @@
+from pydriller.metrics.process.hunks_count import HunksCount
+
+TEST_DATA = [
+    ('https://github.com/ishepard/pydriller', 'scm/git_repository.py', '71e053f61fc5d31b3e31eccd9c79df27c31279bf', '71e053f61fc5d31b3e31eccd9c79df27c31279bf', 8),
```
`'scm/git_repository.py'` should become `str(Path('scm/git_repository.py'))`

stefanodallapalma replied:
Done, thanks.

ishepard commented:
Hi @stefanodallapalma! Thanks for this!
The tests are failing on Windows, my guess is because you didn't use `Path`, but strings. Can you please change them? I put a comment with a suggestion. 

If the CI gives me a green light, I will merge the PR.

As I previously said, it's hard for me to understand if it really works, I will need to add lots of tests and it will require me some time. 
However, if you are ok, from now on I will make you responsible for what's relative to the `process` directory. Hence, all the issues, bugs, and PR that will arrive in the future, I will pass them to you 😄

stefanodallapalma commented:
> Hi @stefanodallapalma! Thanks for this!
> The tests are failing on Windows, my guess is because you didn't use `Path`, but strings. Can you please change them? I put a comment with a suggestion.
> 
> If the CI gives me a green light, I will merge the PR.
> 
> As I previously said, it's hard for me to understand if it really works, I will need to add lots of tests and it will require me some time.

I've  implemented the metrics based on my understanding of the suggested paper (and the related ones). The idea is that developers/researchers can improve them (or propose changes) in the future to provide a common baseline to process those metrics when mining repositories.

> However, if you are ok, from now on I will make you responsible for what's inside the `process` directory. Hence, all the issues, bugs, and PR that will arrive in the future, I will pass them to you

That makes sense to me. Please, go ahead!

ishepard commented:
Yessss, green light! Thanks!
